### PR TITLE
ci: block non-linear merges

### LINKED ISSUE #10903: CI: Reverse merge/updating PR branches with merges can interfere with cherry-picks
Reverse merges into a pull request branch can interfere with some Git operations and lead to unexpected diffs in the history of the master branch.

* Can GitHub block these by default? (Likely: no.)
* Can a GitHub Action as a check on the PR branch detect these and block them? (Likely yes! We can look at the commits between the head & the target and see if there's a merge.)

### PR DESCRIPTION
**Pull request by AaronFriel**
This ensures commit history is linear, enabling customer-owned forks of the pulumi CLI to more easily maintain their fork. Reverse merges into PR branches result in a more complex process for them and for us.

To test this PR, I based my PR branch off an older commit from the target branch and added a merge commit. That resulted in the check failing:
https://github.com/pulumi/pulumi/actions/runs/3416559764/jobs/5686840393

> Checking merge commit efb7be0747f8d19ae77033ba9f2a94865ac15153  for non-linear history
> Main branch parent is: 83c9dfcb9 Merge #11262
> PR branch parents are d9460b9047658ca574d99f0e436190f4d044cbe5
> Checking: d9460b904 Merge remote-tracking branch 'origin/master' into friel/block-reverse-merge
> Error: Non-linear history, PR contains a merge d9460b9047658ca574d99f0e436190f4d044cbe5. Remove this by rebasing on the target.
> Error: Detected non-linear history.
> Error: Process completed with exit code 1.

Fixes #10903

The script used by this lint contains several "tests" and useful diagnostics. Example outputs:

```shell
$ ./scripts/git-linear-history-check.sh f033d9de02a633ad386b09d6dfff810ffe7ddea5
Checking merge commit f033d9de02a633ad386b09d6dfff810ffe7ddea5 for non-linear history
Main branch parent is: 0797f2966 Merge #10817
PR branch parents are 110dd76629cbea137c9b43bf333442765601152c ffbb03c71e715d3dcb93e1f609bc3359d6fbda9e cdf8f20a656ac7e87823b15fe1f5c2384d971a53 396650a814bfa1a64cb1c0b0880e00c20651477b
Checking: 110dd7662 ci: Pin yarn lockfile for security & dependency scanning
Checking: ffbb03c71 ci: Build binary with .exe extension on Windows
Checking: cdf8f20a6 ci: Remove several test skips, check if unnecessary
Checking: 396650a81 ci: Re-enable Windows tests with temp dir
✅ Commit history is linear.
```

```shell
$ ./scripts/git-linear-history-check.sh 0f3e53688fe04ec18180ba87f6915c454023ddf9 
Checking merge commit 0f3e53688fe04ec18180ba87f6915c454023ddf9 for non-linear history
Main branch parent is: bc704afa5 Merge #10703 #10717
PR branch parents are 9065d7caf06ecb970e98a7fba518745038069d9d 22f2989c196daf984d062e75ea18f45ba5e39192 9f5ec4a992c7b1c393cc8518db2a36039db7c6a6
Checking: 9065d7caf refactored defaultServiceLoop into its own method
::error::Non-linear history, PR contains a merge fa09da62271e1cbc9ca6f9a37d6458fcc6d4d90d. Remove this by rebasing on the target.
::error::Non-linear history, PR contains a merge 536f3d6480601fcc541f5cfe7b01124f7fe61f4c. Remove this by rebasing on the target.
Checking: 22f2989c1 ci: Fix package parallelism assignment
Checking: 9f5ec4a99 Add missing `ProgramTestOptions` overrides in `With`
::error::Detected non-linear history.
```

The `::error::` messages should appear in GitHub Actions logs at the top level (the workflow level) as well as in the detailed output of the action.

### CODE REVIEW

File: scripts/git-linear-history-check.sh [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#
+
+# ## Non-linear merge example
+#
+# Merge commit hash: fc7c341c38c006f96ac288ebfcf5ce18b8e31a48
+#
+# PR commits: https://github.com/pulumi/pulumi/pull/11095/commits
+#
+# This commit is a single merge commit produced by `bors`. The 2nd parent of the commit is the PR
+# commit HEAD, and the commit history shos . This script should log errors and exit with a 1 on this
```
Comment (Frassle) [inline scripts/git-linear-history-check.sh:22]: "shos ."?

File: scripts/git-linear-history-check.sh [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+TARGET_BRANCH_HEAD="${PARENTS[0]}"
+>&2 echo "Main branch parent is: $(git log --oneline -n 1 "${TARGET_BRANCH_HEAD}")"
+# Subsequent parents are from PR branches:
+PR_BRANCH_HEADS=( "${PARENTS[@]:1}" )
+>&2 echo "PR branch parents are ${PR_BRANCH_HEADS[*]}"
+
+HAS_MERGE_COMMIT=false
+for PR_COMMIT in "${PR_BRANCH_HEADS[@]}"; do
+  >&2 echo "Checking: $(git log --oneline -n 1 "${PR_COMMIT}")"
+  # MERGE_BASE=$(git merge-base )
```
Comment (Frassle) [inline scripts/git-linear-history-check.sh:101]: Unused?

File: scripts/git-linear-history-check.sh (hunk 1 above)
Comment (AaronFriel) [reply scripts/git-linear-history-check.sh:22]: Weird! I used an extension to rewrap this after typing it up...

**Review by kpitzen — APPROVED** [review]

**Comment by AaronFriel** [comment]
bors merge

**Comment by AaronFriel** [comment]
bors merge

**Comment by AaronFriel** [comment]
bors retry

**Pull request merged**